dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/task/statemachine/AbstractTaskStateAction.java

```diff
@@ -0,0 +1,310 @@
+                                                          final TaskSuccessLifecycleEvent taskSuccessEvent) {
+        final TaskInstance taskInstance = taskExecutionRunnable.getTaskInstance();
+        taskInstance.setState(TaskExecutionStatus.SUCCESS);
+        taskInstance.setEndTime(taskSuccessEvent.getEndTime());
+        taskInstance.setVarPool(taskSuccessEvent.getVarPool());
+        taskInstanceDao.updateById(taskInstance);
+    }
+
+    /**
+     * Failover task.
+     * <p> Will try to take over the task from remote executor, if take-over success, the task has no effect.
+     * <p> If the take-over fails, will generate a failover task-instance and mark the task instance status to {@link TaskExecutionStatus#NEED_FAULT_TOLERANCE}.
+     */
+    protected void failoverTask(final ITaskExecutionRunnable taskExecutionRunnable) {
+        if (!taskExecutionRunnable.isTaskInstanceInitialized()) {
+            throw new IllegalStateException("The task instance hasn't been initialized, cannot take over the task");
+        }
+        if (takeOverTask(taskExecutionRunnable)) {
+            log.info("Failover task success, the task {} has been taken-over", taskExecutionRunnable.getName());
+            return;
+        }
+        taskExecutionRunnable.initializeFailoverTaskInstance();
+        tryToDispatchTask(taskExecutionRunnable);
+        log.info("Failover task success, the task {} has been resubmitted.", taskExecutionRunnable.getName());
+    }
+
+    private boolean takeOverTask(final ITaskExecutionRunnable taskExecutionRunnable) {
+        if (!taskExecutionRunnable.isTaskInstanceInitialized()) {
+            log.debug("Task: {} doesn't initialized yet, cannot take over the task", taskExecutionRunnable.getName());
+            return false;
+        }
+        if (TaskTypeUtils.isLogicTask(taskExecutionRunnable.getTaskInstance().getTaskType())) {
+            return false;
+        }
+        if (StringUtils.isEmpty(taskExecutionRunnable.getTaskInstance().getHost())) {
+            log.debug("Task: {} host is empty, cannot take over the task", taskExecutionRunnable.getName());
+            return false;
+        }
+        try {
+            final TakeOverTaskRequest takeOverTaskRequest = TakeOverTaskRequest.builder()
+                    .taskInstanceId(taskExecutionRunnable.getTaskInstance().getId())
+                    .workflowHost(masterConfig.getMasterAddress())
+                    .build();
+            final TakeOverTaskResponse takeOverTaskResponse = SingletonJdkDynamicRpcClientProxyFactory
+                    .withService(ITaskInstanceOperator.class)
+                    .withHost(taskExecutionRunnable.getTaskInstance().getHost())
+                    .takeOverTask(takeOverTaskRequest);
+            return takeOverTaskResponse.isSuccess();
+        } catch (Exception ex) {
+            log.info("Take over task: {} failed", taskExecutionRunnable.getName(), ex);
```

Comment:
```suggestion
            log.error("Take over task: {} failed", taskExecutionRunnable.getName(), ex);
```

Reply:
I changed to warn since if the worker is not alive, there will be exceptions here.